[Wan] 미션 3 : 체스판 초기화

MyungWanPark opened the pull request:
안녕하세요 백엔드 반의 Wan 입니다! 
제가 구현한 '미션3: 체스판 초기화' 의 특이사항에 대해 잠깐 소개해 드리겠습니다~!

1. main/java/chess/ChessMain.java
- 메인 메소드에서 스캐너로 명령을 입력 받습니다. 
- 메인 메소드 안에 무한 반복문을 돌려서 명령대로 실행하는 것을 구현하였습니다 

2. main/java/chess/Board.java
- step2 까지는 흑색과 백색 폰으로 나누지 않았지만 initialize 메소드에서 각각의 폰을 구분해서 PawnList에 저장해야 하므로 모든 메소드를 흑과 백으로 나눠서 구현하였습니다.
- getWhitePawnsResult()에서는 StringBuilder로 현재 들어간 흰색폰의 갯수만큼 흰색폰의 이니셜을 합쳐서 return으로 보냈습니다. 
- getDotOnBoard() 이 부분은 사실 sb.append("........")을 하려고 했으나 참고코드를 보고 메소드로 표현하는게 더 멋져보여서 
메소드로 구현했습니다.

3. main/java/pieces/Pawn.java
- 체스판에 표기할 폰의 이니셜을 흰색은 소문자 p 검정색은 대문자 P로 나타내기 위해 상수를 사용하였습니다.

4. test/java/chess/BoardTest.java
- initialize()은 Board 클래스에서 폰을 초기에 8개를 생성하고 이것이 폰 배열에 넣어지는 지 테스트하기 위한 메소드입니다.
- getWhitePawnsResult()는 흰색 폰을 체스판에 다 배치하였을때 표기되는 이니셜이 pppppppp이 맞는지 테스트 합니다.
- printBoard()는 일단 폰으로 초기화가 된 후에 화면에 그 모습이 출력이 잘 되었는지 확인하는 테스트 메소드 입니다.

이번에 구현하면서 궁금한 점이 1가지 생겼습니다~! 
test/java/chess/BoardTest.java 에서 printBoard() 메소드는 출력 테스트를 편하게 하기위해 테스트코드로 작성하였습니다. 하지만 테스트 메소드임에도 불구하고 assertThat.isEqualTo를 사용하지 않았는데 그래도 관습상 괜찮을까요?

부족한 점이 많은 코드입니다ㅜㅜ 따끔한 조언 부탁드릴께요~! 
바쁘신 와중에 긴글 읽어주시고 코드리뷰해주셔서 감사드립니다!

MyungWanPark committed:
feat: Remove enum for readability and add constant

Remove enum because it doesn't make readability better. Instead, I put
constant.

MyungWanPark committed:
feat: Change the name of instance variable

Change the name of instance variable pawn1 to whitePawn

MyungWanPark committed:
feat: Remove constructor initializing pawnList

Remove constructor initalizing pawnList in Board.java for readability

MyungWanPark committed:
feat: Remove repeated part in PawnTest.java

Split the method named createPawn to createPawn and verifyPawn for
readability

MyungWanPark committed:
feat: add final keyword on parameter

MyungWanPark committed:
feat: Add representation of pawn

Add representation of pawn in Pawn.java and change regardding method and
test method in PawnTest.java and BoardTest.java

MyungWanPark committed:
Add: add initialize method in BoardTest and Board.java

MyungWanPark committed:
Add: add annotation in initialize() in boardtest.java

MyungWanPark committed:
Add: add initialize method in BoardTest.java

add initialize method in BoardTest.java and Test method regardding it

MyungWanPark committed:
feat: Split all the method to white and black

Split all the method and variable to white pawn and black pawn version

MyungWanPark committed:
Feat: Add getWhitePawnResult

Add getWhitePawnResult and getBlackPawnResult by using StringBuilder and
for loop

MyungWanPark committed:
Refactor: Add PAWN_NUMBER && change access modifier

Add constant in Board.java and changed access modifer precisely as
Object to String

MyungWanPark committed:
Add: Add printBoard methhod in Board.java

Add printBoard method in Board.java and test method regarding that in
BoardTest.java

MyungWanPark committed:
Add: Make printBoard() in Board.java

Make printBoard() in Board.java and make test method regarding that

MyungWanPark committed:
Add: Add main method in ChessMain.java

MyungWanPark committed:
feat: Make main method in ChessMain.java

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,79 @@
+
+    void initialize() {
+        for(int i = 0; i < PAWN_NUMBER ; i++){
+            Pawn whitePawn = new Pawn(Pawn.WHITE,Pawn.WHITE_REPRESENTATION);
+            whitePawnAdd(whitePawn);
```
생성자를 보고 있으면 하얀 폰이 검정색을 갖도록 생성가능할 것 같은데 이게 맞을까요?

MyungWanPark replied:
앗, 제가 확실히 이해를 못해서 그런데,  흰색 폰이 검정색 폰의 이니셜을 가질 수도 있다는 말씀이신건가요?? :)

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,79 @@
+        return sb.toString();
+    }
+
+    String getBlackPawnsResult() {
```
이 메소드 이름 요구사항에 있는 거 아니죠? 저는 이 이름이 맘에 들지 않는데 너무 많은 분들이 같은 이름을 써서...
Result가 추상적이로 코드를 보지 않으면 뭔지 잘 모르겠어요. 그냥 getBlackPawnRepresentations() 어때요?

MyungWanPark replied:
네, 요구사항에는 없구 루카스 예시에 이렇게 이름이 지어져있어서 그냥 수용해버렸네요! 앞으로 명확하게 메소드 이름을 지을 수 있도록 노력하겠습니다! 조언 감사해요!!

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,79 @@
+    String getDotOnBoard(){
+        return "........";
+    }
+    void printBoard(){
```
위에 공백 한 줄

MyungWanPark replied:
앗, 수정하겠습니다! 꼼꼼히 봐주셔서 감사합니다

honux77 replied:
private을 놓쳤는데 ㅋㅋ 꼼꼼하지 않네요.

MyungWanPark replied:
앗, 수정하겠습니다!

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,79 @@
+    void printBoard(){
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getWhitePawnsResult()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getBlackPawnsResult()+"\n");
+        sb.append(getDotOnBoard());
```
'\n' 중복이 많이 발생하는군요. 아마 그게 다음 미션에 있을 거에요. 일단 패스~

MyungWanPark replied:
앗, '\n'의 중복에 대한 해결책을 떠올리지 못해서 더 고민하지 않고 넘어가버렸네요 ㅠㅠ 다음부터 중복에 대해 좀 더 고민한 흔적을 남기겠습니다! 감사합니다.

honux77 commented on src/main/java/pieces/Pawn.java:
```diff
@@ -1,18 +1,29 @@
     }
 
-    public Pawn(Color color) {
+    public Pawn(final String color, final char representation) {
```
color만 받도록 수정

MyungWanPark replied:
최대한 사용자 입력의 오류를 받지 않도록 수정하겠습니다~! 조언 감사합니다~!

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -17,15 +16,36 @@ void createBoard(){
     @Test
     @DisplayName("보드를 생성한 후 폰 테스트 한다.")
```
뭘 테스트하는지 구체적으로 쓰지 않으면 의미가 없어요.

MyungWanPark replied:
앗, DisplayName의 목적을 생각하지 않고 작성했던 것 같습니다. 구체적으로 무엇을 테스트하는지 명확하게 작성하겠습니다. 조언 감사합니다.

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -17,15 +16,36 @@ void createBoard(){
+    }
+
+    @Test
+    @DisplayName("보드판을 초기화 합니다")
```
여기도 마찬가지

MyungWanPark replied:
넵, 여기도 테스트하는 목적을 생각해서 구체적으로 작성하겠습니다!

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -17,15 +16,36 @@ void createBoard(){
+    }
+
+    @Test
+    @DisplayName("보드판을 출력합니다")
```
assert가 없으면 사실 테스트로 의미는 없는데요. 그래도 내가 눈으로 확인하기 위해서 그럴 수 있겠죠.

honux77 replied:
눈으로 확인하는 것도 역시 테스트로서의 가치는 없겠네요.

MyungWanPark replied:
아하 테스트로써 가치가 없군요..., 혹시 그러면 눈으로 확인하기 위해 출력하는 메소드는 테스트 메소드로 작성하지 않교 다른 추천해주시는 방법이 있을까요??

honux77 replied:
테스트가 잘 되는 코드가 좋은 코드다라는 말도 있어요. 어떻게 하면 좋을까요? 
근데 가치는 없지만 짜는게 안 되는 건 아니니까요. 확인 용도로 쓸 수는 있을 것 같아요.

MyungWanPark replied:
조언해주셨던대로 printBoard()를 StringBuilder을 String으로 합쳐서  return하는 메소드와 이를 출력하는 메소드로 분리해서 테스트를 실시할 수 있었습니다! 감사합니다~!

honux77 reviewed (changes requested):
크게 수정할 것 없이 잘 했어요. 코멘트 보시고 조금 개선해 주세요.

MyungWanPark committed:
feat: Remove representation parameter in Pawn()

Remove representation parameter in Pawn() to prevent from strange input
&& Change argument in createPawn() named addBlackPawnOnBoard from 2 to 1 in BoardTest.java
becuase I split all the method to white and black.

MyungWanPark committed:
Refactor: Rename the name of method

Rename the name of method getWhitePawnsResult to
getWhitePawnRepresentation to express the meaning of method more precisely.
And do the same to getBlackPawnsResult to getBlackPawnsRepresentation.

MyungWanPark committed:
Add: add enter between the method for readability

MyungWanPark committed:
Refactor: Rename DispalyName annotation to clarify the purpose of test method

MyungWanPark committed:
Refactor: Rename DispalyName annotation to clarify the purpose of test method

MyungWanPark committed:
Refactor: Rename DispalyName annotation to clarify the purpose of test method

MyungWanPark commented:
소중한 리뷰 정말 감사합니다!
개선한 코드 올렸습니다~!

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,80 @@
-        pawnList = new ArrayList<>();
+    public static final int PAWN_NUMBER = 8;
+
+    void whitePawnAdd(Pawn whitePawn) {
```
앗 그리고 지금 보니 접근 한정작 없는 메소드들이 많이 있네요. private을 붙여 주세요.

MyungWanPark replied:
private을 붙이면 test클래스에서 접근할 수 없어서 에러가 나는데 그럴경우, private 메소드로 접근하는 public 메소드를 하나 더 만드는게 좋을까요 :?

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -7,22 +7,80 @@
-    public Pawn findPawn(int index) {
-        return pawnList.get(index);
+    void printBoard(){
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getWhitePawnsRepresentation()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getDotOnBoard()+"\n");
+        sb.append(getBlackPawnsRepresentation()+"\n");
+        sb.append(getDotOnBoard());
+
+        System.out.println(sb.toString());
```
이 부분을 String으로 리턴하는 메소드와 출력하는 메소드로 나누면 String 리턴하는 메소드를 테스트할수 있겠죠?

MyungWanPark replied:
오오!! 이부분을 어떻게 테스트 하나 고민하고 있었는데 조언 감사합니다! 바로 수정하겠습니다

honux77 reviewed (changes requested):
몇 가지 놓친 부분이 있었네요.
private 한정자 붙이기랑 String 리턴 메소드 만들기 등 다시 한번 수정해 주세요1

honux77 commented on src/main/java/chess/ChessMain.java:
```diff
@@ -0,0 +1,27 @@
+
+public class ChessMain {
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
```
scanner close()는 언제??

MyungWanPark replied:
앗! 그러네요! Try-with-resource 문으로 감싸서 수정했습니다!!

MyungWanPark committed:
Add: close the scanner by scanner.close()

MyungWanPark committed:
feat: Change printBoard() to getBoard() to test the Board

MyungWanPark committed:
Refactor: Put try-with-resource in ChessMain.java

MyungWanPark commented:
조언해주셔서 감사합니다. 수정하였습니다! 

한가지 궁금한 점이 Board.java 에 있는 메소드들에 private을 붙이게 되면 test폴더 내에 있는 클래스에서 접근이 불가능한데 private메소드로 만들고 이에 접근하는 public 메소드를 만드는 식으로 바꾸는게 좋을까요?

honux77 commented:
private 메소드는 테스트를 할 수 없고  public 메소드를 통해서 간접적으로 테스트를 하는게 정상입니다. private 메소드를 테스트하기 위해서 억제로 테스트를 위한 public 메소드를 만든다면 그것도 어딘가 잘못된 것이 아닌지 고민해 봐야 할 것 같습니다. 수고하셨어요!
